### PR TITLE
Cache artifact root for Databricks artifact repository

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Every time an MLflow artifact is uploaded / downloaded on Databricks, an instance of `DatabricksArtifactRepository` is created. Creation of `DatabricksArtifactRepository` makes a network request to the `GetRun` API to fetch the artifact location associated with the target run. On Databricks, the artifact location is immutable, and the run ID is globally unique. Accordingly, this PR maintains a global cache of artifact locations associated with MLflow runs on Databricks, reducing the volume of `GetRun` requests when multiple artifacts are uploaded / download to / from the same run.

## How is this patch tested?

- Unit tests
- Manual tests on AWS / Azure Databricks ensuring that artifact uploads / downloads succeed, including multiple U/Ds for the same run and U/Ds across runs

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/artifact/databricks_artifact_repo.py [hunk 1]
```diff
@@ -122,9 +125,23 @@ def _call_endpoint(self, service, api, json_body):
...
+        if cached_root is not None:
+            return cached_root
+        else:
+            json_body = message_to_json(GetRun(run_id=run_id))
+            run_response = self._call_endpoint(MlflowService, GetRun, json_body)
+            artifact_root = run_response.run.info.artifact_uri
+
+            # Cache the artifact root to avoid a future network call, removing the oldest
+            # entry in the cache if there are too many elements
+            if len(DatabricksArtifactRepository._artifact_roots_cache) > 1024:
```
Comment (arjundc-db) [inline mlflow/store/artifact/databricks_artifact_repo.py:140]: Do you wanna add a small unit test here for the sake of completeness?

**Review by arjundc-db — APPROVED** [review]
LGTM, thanks @dbczumar

File: mlflow/store/artifact/artifact_repo.py [hunk 2]
```diff
@@ -14,6 +14,9 @@ class ArtifactRepository:
     """
     Abstract artifact repo that defines how to upload (log) and download potentially large
     artifacts from different storage backends.
+
+    Artifact repositories are cached per run ID upon instantiation. It is therefore not
+    recommended to alter the artifact root before a run has terminated.
```
Comment (dbczumar) [inline mlflow/store/artifact/artifact_repo.py:19]: @apurva-koti Can we move this to a location that appears in the public / generated docs on mlflow.org? E.g. we could add this information to https://mlflow.org/docs/latest/tracking.html#artifact-stores under the current text:

```
The artifact store is a location suitable for large data (such as an S3 bucket or shared NFS file system) and is where clients log their artifact output (for example, models). artifact_location is a property recorded on mlflow.entities.Experiment for default location to store artifacts for all runs in this experiment. Additional, artifact_uri is a property on mlflow.entities.RunInfo to indicate location where all artifacts for this run are stored.
```

We could add something like the following:

```
The MLflow client caches artifact location information on a per-run basis. It is therefore not recommended to alter the artifact location before a run has terminated.
```

File: mlflow/tracking/_tracking_service/client.py [hunk 3]
```diff
@@ -255,11 +258,23 @@ def _record_logged_model(self, run_id, mlflow_model):
         self.store.record_logged_model(run_id, mlflow_model)
 
     def _get_artifact_repo(self, run_id):
```
Comment (dbczumar) [inline mlflow/tracking/_tracking_service/client.py:260]: @apurva-koti Can we add a unit test here to ensure that the same object is fetched every time `_get_artifact_repo` is called with a particular run ID?

File: mlflow/store/artifact/artifact_repo.py (hunk 2 above)
Comment (apurva-koti) [reply mlflow/store/artifact/artifact_repo.py:19]: Yep sounds good

Comment (apurva-koti) [reply mlflow/store/artifact/artifact_repo.py:19]: Done!

File: mlflow/tracking/_tracking_service/client.py (hunk 3 above)
Comment (apurva-koti) [reply mlflow/tracking/_tracking_service/client.py:260]: Done!

**Comment by apurva-koti** [comment]
Verifies the new unit test fails on master.

**Comment by dbczumar** [comment]
@apurva-koti Thanks for the changes! They look great! Merging!

**Pull request merged**